### PR TITLE
AuditLog parser

### PR DESCRIPTION
**Pull request by robbiezhang**
Create auditlog_lib.rb as a utility class for audit log related operations.
Create parser_auditlog.rb as FluentD parser plugin to convert the audit log into json object
Add unit tests for the utility class
@Microsoft/omsagent-devs

### CODE REVIEW

**Comment by robbiezhang** [comment]
Ping @Microsoft/omsagent-devs

File: source/code/plugins/auditlog_lib.rb [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+          return USER_ID_FIELDS_MAPPING[name], get_user_name_by_id(value.to_i)
+        end
+      
+        if GROUP_ID_FIELDS_MAPPING.has_key? name
+          return GROUP_ID_FIELDS_MAPPING[name], get_group_name_by_id(value.to_i)
+        end
+      rescue => e
+        # type doesn't match
+        @error_handler.log_error("Unable to translate the field: '#{name}': '#{value}'. Message: #{e}")
+      end
```
Comment (niroyb) [inline source/code/plugins/auditlog_lib.rb]: What happens if none of the mappings have the "name" key?

File: source/code/plugins/auditlog_lib.rb [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    end
+    
+    def get_typed_value(name, value)
+      # determine the type of the value by name and convert
+      return value
```
Comment (niroyb) [inline source/code/plugins/auditlog_lib.rb]: This doesn't do anything

File: source/code/plugins/auditlog_lib.rb [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    end
+    
+    def get_process_name_by_id(pid)
+      begin
+        return IO.read("/proc/#{pid}/cmdline").tr("\000", ' ').strip
```
Comment (niroyb) [inline source/code/plugins/auditlog_lib.rb]: `\0` and `\000` are equivalent. The second form is the octal notation. I think most people would prefer the first.

File: source/code/plugins/auditlog_lib.rb [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+      rescue
+        # Process terminated
+      end
+    end
+    
+    def get_user_name_by_id(uid)
+      return 'unset' if uid == 4294967295  # stand for 0xffffffff
```
Comment (niroyb) [inline source/code/plugins/auditlog_lib.rb]: You can use `0xff` in ruby.

File: source/code/plugins/auditlog_lib.rb [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        @error_handler.log_error("Unable to parse the line #{e}")
+      end
+
+      return time, data
+    end
+    
+
+    private
+
+    def remove_quotations(value)
```
Comment (niroyb) [inline source/code/plugins/auditlog_lib.rb]: If you want to remove all quotations consider using `String.tr` otherwise rename to strip_quotations.

**Comment by niroyb** [comment]
Signed off

File: source/code/plugins/auditlog_lib.rb (hunk 1 above)
Comment (robbiezhang) [inline source/code/plugins/auditlog_lib.rb]: The column won't be translated if there is no mapping, and no extra property will be added.

File: source/code/plugins/auditlog_lib.rb (hunk 2 above)
Comment (robbiezhang) [inline source/code/plugins/auditlog_lib.rb]: We don't convert the type for now.

File: source/code/plugins/auditlog_lib.rb (hunk 5 above)
Comment (robbiezhang) [inline source/code/plugins/auditlog_lib.rb]: Good point. I just want to strip the leading and ending quotes

File: source/code/plugins/auditlog_lib.rb (hunk 4 above)
Comment (robbiezhang) [inline source/code/plugins/auditlog_lib.rb]: Good point

File: source/code/plugins/auditlog_lib.rb (hunk 3 above)
Comment (robbiezhang) [inline source/code/plugins/auditlog_lib.rb]: Updated

**Pull request merged**